umbraco/Umbraco-CMS
# V9: Autogeneration of AppSettings JSON schema and Merge the official + Umbraco

### Description
This PRs is the result of a Cycle hack.
We have added a new project `JsonSchema` that uses [NJsonSchema](https://github.com/RicoSuter/NJsonSchema) to generate the schema from Models.

The project creates some dummy models to simulate the `AppSettings:Umbraco:CMS`, `AppSettings:Umbraco:Forms` and `AppSettings:Umbraco:Deploy` part of the configuration paths.

It copies the XML docs from forms from the nuget cache to use inside the `JsonSchema` project and use the docs as information in the generated JSON Schema.

This PR also updates the `AppSettings*.json` files to point to this newly generated schema. To avoid the issue of not showing all non-Umbraco generate IntelliSense for `AppSettings*.json`, we merge the official [AppSettings JsonSchema](https://json.schemastore.org/appsettings.json) into our generated schema.

Futhermore, this PR also contains some refactoring to the way we register configurations using an `UmbracoOptionsAttribute`.

### Next steps
Some ideas for the future
- Add dependency to Deploy, as soon as a nightly is available and generate schema for these.

### Test
- Ensure the generated artifact [`Umbraco.Cms.StaticAssets.9.0.0-preview*.nupkg`](https://dev.azure.com/umbraco/_apis/resources/Containers/18560616/nupkg?itemPath=nupkg%2FUmbraco.Cms.StaticAssets.9.0.0-preview20210625.66532.nupkg) contains the new schema file `content/umbraco/config/appsettings-schema.json`.
- Ensure the generated artifact  [`Umbraco.Templates.9.0.0-preview*.nupkg`](https://dev.azure.com/umbraco/_apis/resources/Containers/18560616/nupkg?itemPath=nupkg%2FUmbraco.Templates.9.0.0-preview20210625.66532.nupkg) has this line 
``` "$schema" : "./umbraco/config/appsettings-schema.json",``` 
in the following files:
  - `appsettings.Development.json`
  - `appsettings.json`
